### PR TITLE
✨  feat(logging): make logging work by default (only v3+)

### PR DESCRIPTION
**Pull request by georgettica**
### Description
This PR makes the `kubebuilder init` initialize the default logging in a way that:
- by Default the log level is `debug`
- any flag passed to the operator will change that log level accordingly (see https://sdk.operatorframework.io/docs/building-operators/golang/references/logging/)

### Motivation
to reduce friction and allow an operator developer not have to meddle with the main.go if not required, this will be related to trying inside the Reconcile function to call `r.Log.V(2).Info("message")` and trying to enable it on demand

### Related issues
- #885 related to changing logging levels

### CODE REVIEW

**Comment by camilamacedo86** [comment]
Hi @georgettica, 

Thank you for your contribution.  🥇 
Could you please squash the commits for we are able to move forward with?

**Comment by georgettica** [comment]
do you think this could count as an `help wanted` issue?

**Comment by georgettica** [comment]
@camilamacedo86 rewrote the description and squashed commits

**Comment by georgettica** [comment]
as the test doesn't have an issue with my code and only a timeout will retest it 
/test pull-kubebuilder-e2e-k8s-1-17-0

**Review by camilamacedo86 — APPROVED** [review]
It shows ok for me 👍 
/approve

@joelanford WDYT?

**Comment by georgettica** [comment]
@camilamacedo86 I think there is something off in my local deployment of this.. 

I seem to not be able to configure the log level from the deployment

can you TAL on that whenever you have a cycle?

for now
/hold

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
Hi @georgettica, 

shows that your project has a diff/old version of controller-runtime.  
however, your comment makes me think that would be nice to add a check to ensure this change in kubebuilder/test/e2e/v3/. wDYT could you please add it for we are able to move forward with?

**Comment by camilamacedo86** [comment]
/approve cancel

Let's add a test before get it merged to ensure. WDYT?

**Comment by camilamacedo86** [comment]
> @camilamacedo86 I want to add this change but it is a bit above my level of understanding for now.. can I merge this and pass the autotests to someone else?

The tests would ensure that it works so it usually is part of the PR. However, since it is very small I think that I am ok with. 👍 
/hold cancel

**Review by prafull01 — COMMENTED** [review]
Verified this change locally. It shows OK for me :+1:

**Comment by georgettica** [comment]
@prafull01 can you help with guiding me on this?

> however, your comment makes me think that would be nice to add a check to ensure this change in kubebuilder/test/e2e/v3/. WDYT could you please add it for we are able to move forward with?
I know the change is to be made here https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/e2e_suite.go, but aside of that I wasn't able to figure out how to add my validation

**Comment by prafull01** [comment]
> @prafull01 can you help with guiding me on this?
> 
> > however, your comment makes me think that would be nice to add a check to ensure this change in kubebuilder/test/e2e/v3/. WDYT could you please add it for we are able to move forward with?
> 
> I know the change is to be made here https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/e2e_suite.go, but aside of that I wasn't able to figure out how to add my validation

@georgettica This zap logging changes are recently added to the controller-runtime and is not present in the older versions. Hence, e2e test should have a check which verifies that the log `controller-manager` is generating is following the `log level`, `stacktrace` level etc. The output of logs before this change is something like this:
```
2020-10-22T10:19:04.236+0530	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": ":8080"}
2020-10-22T10:19:04.236+0530	INFO	controller-runtime.builder	Registering a mutating webhook	{"GVK": "crew.testproject.org/v1, Kind=Captain", "path": "/mutate-crew-testproject-org-v1-captain"}
```
The output of controller-manager logs after this change:
```
{"level":"info","ts":1603342108.069236,"logger":"controller-runtime.metrics","msg":"metrics server is starting to listen","addr":":8080"}
{"level":"info","ts":1603342108.069433,"logger":"controller-runtime.builder","msg":"Registering a mutating webhook","GVK":"crew.testproject.org/v1, Kind=Captain","path":"/mutate-crew-testproject-org-v1-captain"}
```
So, verification of this change can happen just after this, in the e2e tests:
```
By("validate the controller-manager pod running as expected")
``` 

@camilamacedo86 Please correct me if there is mistake in my understanding.

**Comment by camilamacedo86** [comment]
Really thank you @prafull01. It is perfect.

@georgettica, 

Ideally, when we do a change we need to ensure its tests as well. However, since @prafull01 help here and manually checks it in https://github.com/kubernetes-sigs/kubebuilder/pull/1721#issuecomment-714301930 and it is very small, if you still strangle to add the tests as he described I think that it is fine we can move forward and to do it a follow-up. But we need to raise an issue at least for that do not get lost.  Let's see what others think as well. So, it is for me: 

/approve

**Review by camilamacedo86 — APPROVED** [review]
The follow up to the tests was tacked in : https://github.com/kubernetes-sigs/kubebuilder/issues/1739

/approve 
/lgtm

**Comment by georgettica** [comment]
Thanks @camilamacedo86 ! I will follow up on it there and see if I can help.push it forward

**Comment by estroz** [comment]
There's an upstream bug that results in `Development: true` not being applied, see https://github.com/kubernetes-sigs/controller-runtime/issues/1228 (fixed in v0.7.0).

**Pull request merged**
